Feature/snapshot weights

martinhjel opened the pull request:
As discussed in #639 

## Changes proposed in this Pull Request
- Add flag to compute snapshot_weightings with timedelta if snapshots are of type pd.DatetimeIndex. Default is False.
- Add argument for selecting snapshot weightings, set default to 1.0.
- Add pandapower and scikit-learn to environment.yaml. Limit pandas version as pandapower fails for version >=2.0 of pandas due to .iteritems(). 
- Add type hints.
- Add pyproject.toml file for project configurations to align with [PEP 518](https://peps.python.org/pep-0518/) (should perhaps have been added in a separate PR..)

## Checklist

- [x] Code changes are sufficiently documented; i.e. new functions contain docstrings and further explanations may be given in `doc`.
- [x] Unit tests for new features were added (if applicable).
- [x] Newly introduced dependencies are added to `environment.yaml`, `environment_docs.yaml` and `setup.py` (if applicable).
- [x] A note for the release notes `doc/release_notes.rst` of the upcoming release is included.
- [x] I consent to the release of this PR's code under the MIT license.

FabianHofmann commented:
Hey @martinhjel! Thanks for your contribution! 

The PR looks nice  :) But it is mixing topics a bit. I am wondering about the pandas limitation. The pandapower repo also uses pandas >= 2.0 in there CI (see https://github.com/e2nIEE/pandapower/actions/runs/5043342658/jobs/9045012640#step:4:53). Did you update to their latest package?

martinhjel commented:
Found out I had pandapower 2.8.0 installed, updated to 2.13, and the `AttributeError: 'Series' object has no attribute 'iteritems'` disappeared. Removed the dependency now.

FabianHofmann commented on doc/release_notes.rst:
```diff
@@ -17,6 +17,10 @@ Upcoming Release
   deprecated but all functionality will continue to be accessible until the next
   major version.
+* The function ``n.set_snapshots()`` now takes two arguments; ``default_snapshot_weightings``
+  that can be used to change the default snapshot weightings, and ``weightings_from_timedelta``
```
```suggestion
  to change the default snapshot weightings, and ``weightings_from_timedelta``
```

FabianHofmann commented on pypsa/components.py:
```diff
@@ -454,34 +455,64 @@ def meta(self, new):
+        default_snapshot_weightings: float
+            The default weight for each snapshot. Defaults to 1.0.
+        weightings_from_timedelta: bool
+            Use the timedelta of `snapshots` as `snapshot_weightings` if `snapshots` is of type `pd.DatetimeIndex`.  Defaults to False.
```
```suggestion
            Wheter to use the timedelta of `snapshots` as `snapshot_weightings` if `snapshots` is of type `pd.DatetimeIndex`.  Defaults to False.
```

FabianHofmann reviewed (changes requested):
Just some points regarding the docs

martinhjel commented:
@FabianHofmann, is there anything else I have to do to complete this PR?
